Update calculateFilterExpression and fix T875683, T884232, T892205 (#2335)
DevExpress/devextreme-documentation: api-reference/_hidden/GridBaseColumn/calculateFilterExpression.md

---
---
##### shortDescription

> Specifies the column's custom rules to filter data. Does not apply if you set the column's [headerFilter]({basewidgetpath}//Configuration/columns/headerFilter/).[dataSource]({basewidgetpath}/Configuration/columns/headerFilter/#dataSource).

Reviewer: The calculateFilterExpression callback is not called only if headerFilter.dataSource contains items where value field is filter expression (array) and for header filter only. See the 
https://js.devexpress.com/Demos/WidgetsGallery/Demo/DataGrid/Filtering/jQuery/Light/
For OrderDate column calculateFilterExpression is called. For SaleAmount column it is not called if select values in header filter.

---
The **calculateFilterExpression** function must return a [filter expression](/concepts/70%20Data%20Binding/5%20Data%20Layer/2%20Reading%20Data/15%20Filtering '/Documentation/Guide/Data_Binding/Data_Layer/#Reading_Data/Filtering'). The simplest filter expression has the following format:

    [selector, comparisonOperator, filterValue]

- `selector`      
A [dataField]({basewidgetpath}/Configuration/columns/#dataField) or function that returns column values. Pass `this.calculateCellValue` if your column contains [calculated values](/api-reference/_hidden/GridBaseColumn/calculateCellValue.md '{basewidgetpath}/Configuration/columns/#calculateCellValue').

- `comparisonOperator`       
One of the following operators: *"=", "<>", ">", ">=", "<", "<=", "startswith", "endswith", "contains", "notcontains"*.     

- `filterValue`        
A user input value. Values from the `selector` are compared to this value.

In the following code, the **calculateFilterExpression** function implements an exclusive *"between"* operation. The function overrides the default inclusive implementation. Note that for the *"between"* operation, the filter expression has a different format:

>     [ [selector, ">=", startValue], "and", [selector, "<=", endValue] ]

Reviewer: For endValue DataGrid generates "<" instead of "<=" with endValue correction:
[ [selector, ">=", startValue], "and", [selector, "<", endValue + 1] ]

---
---
##### shortDescription
Specifies the column's custom rules to filter data.

##### param(filterValue): any
A user input value.             
Contains an array if the `selectedFilterOperation` is one of the following: *"between"*, *"anyof"*, *"noneof"*.

##### param(selectedFilterOperation): String
A selected filter operation.

##### param(target): String
A UI element used to filter data.     
Possible values: [*"filterRow"*]({basewidgetpath}/Configuration/filterRow/), [*"headerFilter"*]({basewidgetpath}/Configuration/headerFilter/), [*"filterBuilder"*]({basewidgetpath}/Configuration/#filterBuilder), [*"search"*]({basewidgetpath}/Configuration/searchPanel/).

##### return: Filter expression
A filter expression.          
If you filter data [remotely]({basewidgetpath}/Configuration/remoteOperations/), the filter expression must not contain functions.

---

> The **calculateFilterExpression** function must return a [filter expression](/concepts/70%20Data%20Binding/5%20Data%20Layer/2%20Reading%20Data/15%20Filtering '/Documentation/Guide/Data_Binding/Data_Layer/#Reading_Data/Filtering'). The simplest filter expression has the following format:

Reviewer 2: ```suggestion
The **calculateFilterExpression** function should return a [filter expression](/concepts/70%20Data%20Binding/5%20Data%20Layer/2%20Reading%20Data/15%20Filtering '/Documentation/Guide/Data_Binding/Data_Layer/#Reading_Data/Filtering'). A basic filter expression has the following format:
```

    [selector, comparisonOperator, filterValue]

- `selector`      
A [dataField]({basewidgetpath}/Configuration/columns/#dataField) or function that returns column values. Pass `this.calculateCellValue` if your column contains [calculated values](/api-reference/_hidden/GridBaseColumn/calculateCellValue.md '{basewidgetpath}/Configuration/columns/#calculateCellValue').

- `comparisonOperator`       
One of the following operators: *"=", "<>", ">", ">=", "<", "<=", "startswith", "endswith", "contains", "notcontains"*.     

- `filterValue`        
A user input value. Values from the `selector` are compared to this value.

A filter expression for the *"between"* operation has a different format:

    [ [selector, ">=", startValue], "and", [selector, "<=", endValue] ]

> The default *"between"* implementation is inclusive (filter results include the boundary values). In the following code, the **calculateFilterExpression** function implements an exclusive *"between"* operation:

Reviewer 2: Why `an exclusive`?
